[BUG] remove cursor twice when task task failure recovery

**Describe the bug**
Both the `close()` and `cancel()` method in `FlinkPulsarSource` would be invoked when encountering task failure. The order of calling should be `close() -> cancel()`, however in method `close()`, `FlinkPulsarSource` has already invoked `cancel()` and then close the `PulsarMetadataReader` so that when Flink `Task#doRun()` invoke the `cancel()` method we can receive a `Client instance has been closed` error.

**To Reproduce**
Steps to reproduce the behavior:
1. Prepare a flink program
```java
env
        .addSource(new FlinkPulsarSource<>(serviceUrl, adminUrl, new SimpleStringSchema(), props))
        .addSink(new RichSinkFunction<String>() {
            @Override
            public void invoke(String value, Context context) throws Exception {
                throw new Exception("Simulated Sink Exception");
            }
        });
see.execute("PulsarConnectorNormal");
```
2. See error
```

2021-01-14 15:30:16,991 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Removing subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-4
2021-01-14 15:30:17,000 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Successfully removed subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-4
2021-01-14 15:30:17,000 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Removing subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-7
2021-01-14 15:30:17,011 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Successfully removed subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-7
2021-01-14 15:30:17,011 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Removing subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-1
2021-01-14 15:30:17,021 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Successfully removed subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-1
2021-01-14 15:30:17,029 WARN  org.apache.flink.runtime.taskmanager.Task                    - Source: Custom Source -> Filter -> Sink: Unnamed (2/3)#0 (b00537ef7a1976aa11cff5c4558fb590) switched from RUNNING to FAILED.
java.lang.Exception: Simulated Sink Exception
    at sg.XX.flink.PulsarConnectorFrequentlyRestart$1.invoke(PulsarConnectorFrequentlyRestart.java:53) ~[XX-dev-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
    at sg.XX.flink.PulsarConnectorFrequentlyRestart$1.invoke(PulsarConnectorFrequentlyRestart.java:50) ~[XX-dev-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
    at org.apache.flink.streaming.api.operators.StreamSink.processElement(StreamSink.java:56) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:71) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:46) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:26) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:52) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:30) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.StreamFilter.processElement(StreamFilter.java:40) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:71) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:46) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:26) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:52) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:30) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollectWithTimestamp(StreamSourceContexts.java:310) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collectWithTimestamp(StreamSourceContexts.java:409) ~[flink-dist_2.12-1.12.0.jar:1.12.0]
    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.emitRecordsWithTimestamps(PulsarFetcher.java:380) ~[XX-dev-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
    at org.apache.flink.streaming.connectors.pulsar.internal.ReaderThread.emitRecord(ReaderThread.java:206) ~[XX-dev-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
    at org.apache.flink.streaming.connectors.pulsar.internal.ReaderThread.run(ReaderThread.java:104) ~[XX-dev-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
2021-01-14 15:30:17,030 INFO  org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader - Removing subscription flink-pulsar-356c8a2c-be80-48f6-9f78-0807d9b1a888 from topic XXX-partition-4
2021-01-14 15:30:17,031 ERROR org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSource - Failed to cancel the Pulsar Fetcher java.lang.RuntimeException: Failed to remove cursor for TopicRange{topic=XXX-partition-4, key-range=SerializableRange{range=[0, 65535]}}
    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.removeCursor(PulsarMetadataReader.java:257)
    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.cancel(PulsarFetcher.java:396)
    at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSource.cancel(FlinkPulsarSource.java:682)
    at org.apache.flink.streaming.api.operators.StreamSource.cancel(StreamSource.java:147)
    at org.apache.flink.streaming.runtime.tasks.SourceStreamTask.cancelTask(SourceStreamTask.java:153)
    at org.apache.flink.streaming.runtime.tasks.StreamTask.cancel(StreamTask.java:669)
    at org.apache.flink.runtime.taskmanager.Task.cancelInvokable(Task.java:1289)
    at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:792)
    at org.apache.flink.runtime.taskmanager.Task.run(Task.java:547)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Client instance has been closed.
    at org.apache.pulsar.shade.org.glassfish.jersey.internal.guava.Preconditions.checkState(Preconditions.java:169)
    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyClient.checkNotClosed(JerseyClient.java:257)
    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyWebTarget.checkNotClosed(JerseyWebTarget.java:116)
    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyWebTarget.path(JerseyWebTarget.java:127)
    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyWebTarget.path(JerseyWebTarget.java:36)
    at org.apache.pulsar.client.admin.internal.TopicsImpl.topicPath(TopicsImpl.java:1368)
    at org.apache.pulsar.client.admin.internal.TopicsImpl.deleteSubscriptionAsync(TopicsImpl.java:860)
    at org.apache.pulsar.client.admin.internal.TopicsImpl.deleteSubscriptionAsync(TopicsImpl.java:853)
    at org.apache.pulsar.client.admin.internal.TopicsImpl.deleteSubscription(TopicsImpl.java:826)
    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.removeCursor(PulsarMetadataReader.java:250)
    ... 9 more
```

**Expected behavior**
`metadataReader.removeCursor(topics)` should be only executed once to avoid abundant exceptions in taskmanager.

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarFetcher.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarFetcher.java
@@ -393,7 +393,14 @@ public class PulsarFetcher<T> {
         Set<TopicRange> topics = subscribedPartitionStates.stream()
                 .map(PulsarTopicState::getTopicRange).collect(Collectors.toSet());
 
-        metadataReader.removeCursor(topics);
+        try {
+            metadataReader.removeCursor(topics);
+        } catch (PulsarMetadataReader.ClosedException e) {
+            // metadata reader has already close. Do nothing
+        } catch (Exception e) {
+            throw e;
+        }
+
         // make sure the main thread wakes up soon
         unassignedPartitionsQueue.addIfOpen(PoisonState.INSTANCE);
     }

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java
@@ -242,7 +242,11 @@ public class PulsarMetadataReader implements AutoCloseable {
         }
     }
 
-    public void removeCursor(Set<TopicRange> topics) {
+    public void removeCursor(Set<TopicRange> topics) throws ClosedException {
+        if (closed) {
+            throw new ClosedException();
+        }
+
         if (!useExternalSubscription) {
             for (TopicRange topicRange : topics) {
                 try {

